### PR TITLE
[openssl-windows] Added feature option to build openssl with no-pinshared flag

### PR DESCRIPTION
**Pull request by Sofapriester**
**Change in "openssl-windows" portfile**

**Context**
In cases you have an exe project that loads dynamically a DLL via LoadLibrary() and the DLL has OpenSSL statically linked, OpenSSL prevents the DLL from being unloaded dynamically via FreeLibrary().

More details can be found on OpenSSL GitHub:
https://github.com/openssl/openssl/issues/7598
https://github.com/openssl/openssl/pull/7647

The official solution to this scenario is to build OpenSSL with the no-pinshared flag set.

**Problem:**
If you are installing "openssl-windows" via vcpkg, it is not build with this flag.

**Solution**
Because in the most cases this build flag is not needed or even dangerous, I've added an vcpkg feature option to install OpenSSL with no-pinshared optionally.

**Example**
Optionally you can now install openssl-windows like this
openssl-windows[no-pinshared]:x64-windows-static

### CODE REVIEW

File: ports/openssl-windows/portfile.cmake [hunk 1]
```diff
@@ -32,17 +32,24 @@ vcpkg_find_acquire_program(JOM)
```
Comment (NancyLi1013) [inline ports/openssl-windows/portfile.cmake:32]: Could you please update `set(ENV{PATH} "${NASM_EXE_PATH};$ENV{PATH}")` as
`vcpkg_add_to_path("${NASM_EXE_PATH}")`?

File: ports/openssl-windows/portfile.cmake [hunk 2]
```diff
@@ -20,8 +20,8 @@ vcpkg_download_distfile(ARCHIVE
 )
```
Comment (NancyLi1013) [inline ports/openssl-windows/portfile.cmake:21]: Could you please update `set(ENV{PATH} "$ENV{PATH};${PERL_EXE_PATH}")` as
`vcpkg_add_to_path("${PERL_EXE_PATH}")`?

File: ports/openssl-windows/portfile.cmake [hunk 3]
```diff
@@ -73,7 +80,7 @@ if(NOT DEFINED VCPKG_BUILD_TYPE OR VCPKG_BUILD_TYPE STREQUAL "release")
 
     set(OPENSSLDIR_RELEASE ${CURRENT_PACKAGES_DIR})
```
Comment (NancyLi1013) [inline ports/openssl-windows/portfile.cmake:82]: `vcpkg_test_cmake(PACKAGE_NAME OpenSSL MODULE)` 
This function currently doesn't work. Please remove it or comment this out.

Comment (NancyLi1013) [inline ports/openssl-windows/portfile.cmake:82]: Could you please use `vcpkg_replace_string() `to handle the string that needs to replace in a file?

**Review by NancyLi1013 — CHANGES_REQUESTED** [review]

**Comment by NancyLi1013** [comment]
Hi @Sofapriester 
Thanks for this PR.
Have you tested the feature `no-pinshared` and if it can work well on all platforms?
Please make sure that it at least works on x86-windows, x64-windows, x64-windows-static and x64-linux.

**Comment by Sofapriester** [comment]
> Hi @Sofapriester
> Thanks for this PR.
> Have you tested the feature `no-pinshared` and if it can work well on all platforms?
> Please make sure that it at least works on x86-windows, x64-windows, x64-windows-static and x64-linux.

Hi @NancyLi1013 
As this PR only effects openssl-**windows** ports, I think testing its functionality on Windows platforms is enough, isn't it?

File: ports/openssl-windows/portfile.cmake (hunk 2 above)
Comment (Sofapriester) [reply ports/openssl-windows/portfile.cmake:21]: ok

File: ports/openssl-windows/portfile.cmake (hunk 1 above)
Comment (Sofapriester) [reply ports/openssl-windows/portfile.cmake:32]: ok

File: ports/openssl-windows/portfile.cmake (hunk 3 above)
Comment (Sofapriester) [reply ports/openssl-windows/portfile.cmake:82]: I removed it

Comment (Sofapriester) [reply ports/openssl-windows/portfile.cmake:82]: ok

Comment (Sofapriester) [reply ports/openssl-windows/portfile.cmake:82]: It looks like, that the entire string replacement is just useless in this cmake file. They do not hit, as the effected files do not contain the search-strings.

I also see, that there are some .patch files for openssl available, but they are not applied.

Looks like, that there is a lot outdated. But I suggest, that this should not be tackled in this pull-request. Its not really related to it and does not effect the functionality.

Comment (NancyLi1013) [reply ports/openssl-windows/portfile.cmake:82]: Thanks for your above info. In fact, it doesn't matter. We recommend to use new functions to take place of some deprecated functions. Of course, it have no effect on the functionality. We do this is due to some functions can 
simplify the codes and make it more easier to maintain.

**Comment by NancyLi1013** [comment]
Yes, you're right. You only need to test the features on the platforms or triplets that the ports supported. For `openssl-windows`, testing on Windows platform  is enough.

**Comment by Sofapriester** [comment]
> Have you tested the feature `no-pinshared` and if it can work well on all platforms?

**I've tested it on the following configurations:**
openssl-windows:x64-windows-static: Ok
openssl-windows:x86-windows-static: Ok
openssl-windows[no-pinshared]:x64-windows-static: Ok
openssl-windows[no-pinshared]:x86-windows-static: Ok

openssl-windows:x64-windows: Ok
openssl-windows:x86-windows: Ok
openssl-windows[no-pinshared]:x64-windows: Ok
openssl-windows[no-pinshared]:x86-windows: Ok

**Comment by Sofapriester** [comment]
@NancyLi1013 There was suddenly a test failing. Unfortunately I have no idea what happens there. Do you haven an idea?

**Comment by NancyLi1013** [comment]
@Sofapriester 

Thanks for your update for the feature results.
Sorry, I'm not clear about what you mean `There was suddenly a test failing`.  Is the failure on x64-osx in current CI test?

**Comment by Sofapriester** [comment]
@NancyLi1013 
> Sorry, I'm not clear about what you mean `There was suddenly a test failing`. Is the failure on x64-osx in current CI test?

Yes, the failure on CI x64_osx.

**Comment by NancyLi1013** [comment]
@Sofapriester 
Currently the regressions on CI are not related with this PR. So you can ignore it.
I will rerun this PR once these are fixed.

**Comment by Sofapriester** [comment]
Hi @NancyLi1013 ,
I am done with all my changes. Do you need anything else? Or can this pull request be accepted and merged?

Best regards,
Stefan

**Review by NancyLi1013 — APPROVED** [review]

**Comment by NancyLi1013** [comment]
Hi @Sofapriester 
Could you please merge master commits to this PR? 
The current regression caused by `qwt` cannot be resolved by `/azp run`.

```
-- Downloading https://downloads.sourceforge.net/project/qwt/qwt/6.1.3/qwt-6.1.3.zip...
-- Downloading https://downloads.sourceforge.net/project/qwt/qwt/6.1.3/qwt-6.1.3.zip... Failed. Status: 7;"Couldn't connect to server"
CMake Error at scripts/cmake/vcpkg_download_distfile.cmake:175 (message):
      
      Failed to download file.
      If you use a proxy, please set the HTTPS_PROXY and HTTP_PROXY environment
      variables to "https://user:password@your-proxy-ip-address:port/".
      Otherwise, please submit an issue at https://github.com/Microsoft/vcpkg/issues

Call Stack (most recent call first):
  scripts/cmake/vcpkg_from_sourceforge.cmake:114 (vcpkg_download_distfile)
  ports/qwt/portfile.cmake:1 (vcpkg_from_sourceforge)
  scripts/ports.cmake:73 (include)
```

**Comment by NancyLi1013** [comment]
@Sofapriester 
Could you please resolve the conflicts?

**Comment by Sofapriester** [comment]
Hi @NancyLi1013 ,

is there anything left for me to do in order to bring that PR to microsoft:master?

Best regards.

**Comment by NancyLi1013** [comment]
Thanks for your kindly reminder.
LGTM now, nothing needs to be done. Let's just wait for it merged.

**Comment by ras0219-msft** [comment]
Hi @Sofapriester, thanks for the PR!

Could the DLL not simply use the recommended runtime method?
> Applications can suppress running of the atexit() handler at run time by using the OPENSSL_INIT_NO_ATEXIT option to OPENSSL_init_crypto().

(According to my read of the thread, the DLL will need to call `OPENSSL_cleanup()` before unloading as well).

It also looks like there may be residual issues to file on OpenSSL to fix this in a better way that doesn't require dangerous behavior: https://github.com/openssl/openssl/pull/7647#issuecomment-439454231.

Finally, this introduces a "mode" which is not what features are intended to model -- they only correctly model additive APIs/functionality. The best available option for exclusive modes like this would be to look for a variable expected to be set in the triplet (which is included at the top of every portfile, so you can simply use `if(DEFINED OPENSSL_USE_NOPINSHARED)` in the port).

File: ports/openssl-windows/portfile.cmake [hunk 4]
```diff
@@ -44,6 +44,12 @@ set(CONFIGURE_COMMAND ${PERL} Configure
     ${OPENSSL_SHARED}
 )
 
+if(DEFINED OPENSSL_USE_NOPINSHARED)
```
Comment (Sofapriester) [inline ports/openssl-windows/portfile.cmake:47]: @ras0219-msft : Is this how you suggested?

**Comment by Sofapriester** [comment]
Hi @ras0219-msft ,

Thanks for your reply and your recommendation.

**But first to your questions:**
> Could the DLL not simply use the recommended runtime method?
>> Applications can suppress running of the atexit() handler at run time by using the OPENSSL_INIT_NO_ATEXIT option to OPENSSL_init_crypto().

You are absolutely right. This is how you would "avoid" the cleanup automation in openssl and with it the pinning. But in order to have this feature activated and working, you need to have it build with **no-pinshared**. Without having it build like that, you can call `OPENSSL_init_crypto(OPENSSL_INIT_NO_ATEXIT, nullptr)`, but nothing happens (DLL stays pinned until process exit).

> (According to my read of the thread, the DLL will need to call OPENSSL_cleanup() before unloading as well).

Also this is correct, as can be also read [here](https://www.openssl.org/docs/manmaster/man3/OPENSSL_INIT_set_config_file_flags.html).

My suggestion is, that we concentrate in this PR on how to make this official OpenSSL feature available through vcpkg. In my opinion it should be discussed in the OpenSSL project if this OpenSSL feature is good, bad or dangerous.

**Now to your nice suggestion:**
> Finally, this introduces a "mode" which is not what features are intended to model -- they only correctly model additive APIs/functionality. The best available option for exclusive modes like this would be to look for a variable expected to be set in the triplet (which is included at the top of every portfile, so you can simply use if(DEFINED OPENSSL_USE_NOPINSHARED) in the port).

I have actually added this section to the portfile.cmake:
`if(DEFINED OPENSSL_USE_NOPINSHARED)`
`    set(CONFIGURE_OPTIONS ${CONFIGURE_OPTIONS} no-pinshared)`
`endif()`

Now you could create your own triplet like this:
`# build-openssl-nopinshared-triplet.cmake`
`set(OPENSSL_USE_NOPINSHARED ON)`

And call it f.e. like this:
vcpkg.exe install openssl-windows:build-openssl-nopinshared-triplet --overlay-triplets=PATH_TO_MY_TRIPLETS_FOLDER

I have already updated my PR. Could you please check and tell me, if this is how you suggested?

**Review by strega-nil — APPROVED** [review]

File: ports/openssl-windows/portfile.cmake (hunk 4 above)
Comment (ras0219) [reply ports/openssl-windows/portfile.cmake:47]: Yep!

File: ports/openssl-windows/portfile.cmake [hunk 5]
```diff
@@ -74,7 +80,7 @@ if(NOT DEFINED VCPKG_BUILD_TYPE OR VCPKG_BUILD_TYPE STREQUAL "release")
 
     set(OPENSSLDIR_RELEASE ${CURRENT_PACKAGES_DIR})
 
-    message(STATUS "Configure ${TARGET_TRIPLET}-rel")
+    message(STATUS "Configure ${TARGET_TRIPLET}-rel with options ${CONFIGURE_OPTIONS}")
```
Comment (ras0219) [inline ports/openssl-windows/portfile.cmake:83]:
```suggestion
    message(STATUS "Configure ${TARGET_TRIPLET}-rel")
```

**Comment by Sofapriester** [comment]
Hi @ras0219 and @strega-nil ,
I have done all requested and suggested changes.

Is there anything left before this PR could be merged?

Best regards,
Stefan

File: ports/openssl-windows/CONTROL [hunk 6]
```diff
@@ -1,4 +1,4 @@
 Source: openssl-windows
-Version: 1.1.1g
+Version: 1.1.1g-1
```
Comment (NancyLi1013) [inline ports/openssl-windows/CONTROL:2]:
```suggestion
Version: 1.1.1g
Port-Version: 1
```

**Review by Sofapriester — COMMENTED** [review]
Thanks a lot 👍

**Pull request merged**